### DISCUSSION TITLE
How am I supposed to change this default setting

### DISCUSSION THREAD

**Discussion by mrolli** [LunarVim/LunarVim#3952] [Q&A] [answered]
I quite new to the art of configuring neovim using lua and I'm still struggling to get my head around this style that reminds me of the movie Inception.

I know that I can sometimes edit an "exported" table, i.e. `lvim.builtin.lualine.sections.lualine_z = { "progress", "location" }`. I know that such things are found in the `lvim.builtin.$plugin` namespace where a lot of stuff is exposed in order to be change by me. I'm able to make a connection between the code in the lunarvim repo, where defaults are set all the way up to the point where I see how to inject my changes.

But then again, there are situations where I have no clue how I could change a certain default in my own config. As an example, I'm struggling with the following:

I want to have my lsp virtual_text to also show error codes and not only the error string. I identified where the defaults for this thing is defined, namely here, line 67:
https://github.com/LunarVim/LunarVim/blob/d559dfc1cd9d2622393135ad23101e3e73930f30/lua/lvim/lsp/config.lua#L55-L80

The goal is to simply use the same format like for float, so I tried to set virtual_text to a function.

Now, I see, that this above config table is consumed here:
https://github.com/LunarVim/LunarVim/blob/d559dfc1cd9d2622393135ad23101e3e73930f30/lua/lvim/config/init.lua#L29-L30

and later the diagnostics part is injected to nevoim here:
https://github.com/LunarVim/LunarVim/blob/d559dfc1cd9d2622393135ad23101e3e73930f30/lua/lvim/lsp/handlers.lua#L5-L17

I tried different things in my config to overwrite this virtual_text still in the lvim namespace but also later in the vim namespace, with something like the following

```
local my_diag_format = function(d)
  return string.format("(%s) %s", d.code, d.message)
end

???{
   virtual_text = {
    format = my_diag_format,
  },
```

But I as said, how to correctly inject this aka replace the ???. I end up getting errors like "try to set on nil" or bigger errors because of overwriting important stuff or nothing happens and the defaults win, like in the following attempt:

```
vim.diagnostic.config({
  virtual_text = {
    format = my_diag_format,
  },
}

Although the above code was last in my config.lua and doesn't through any error, I assume that lvim overwrites this way later in execution time.

Any help on this? Is there a good place that easy teaches why and how these things are built and how to efficiently identify these extension points?

**Comment by LostInTheLogs** [comment, marked as answer]
lvim.lsp.diagnostics.virtual_text = ...

**Reply by mrolli** [reply]
Apparently I have tried different combination of LHS/RHS and never catched the right one. ;-)
Thanks for "locking" the LHS of the equal sign. Worked now and the solution is even re-using the same format as is set for float:

```
lvim.lsp.diagnostics.virtual_text = {
    format = lvim.lsp.diagnostics.float.format
}
```

Your help is very appreciated!
